### DISCUSSION TITLE
Time Series Exploratory Analysis | Autocorrelation Function (ACF)

### DISCUSSION THREAD

**Discussion by ngupta23** [pycaret/pycaret#1765] [FAQs]
# Background 

Exploratory Analysis is extremely important in any data analysis project but more so in Time Series Analysis. One of the tools at our disposal is the Autocorrelation function or ACF. ACF tells us how a data set is correlated to lagged version of itself. The shape and magnitude of the ACF plot can reveal a lot about the underlying process generating the time series data. Let's see how we can use `pycaret` to diagnose these characteristics.

# How to analyze ACF in `pycaret`?

Luckily, `pycaret` provides a very handy way to analyze ACF plots using interactive plots. Let's look at a few examples

## Common Setup

```python
#### Install library (one time) ----
!pip install pycaret-ts-alpha -U

#### Import libraries ----
from pycaret.datasets import get_data
from pycaret.internal.pycaret_experiment import TimeSeriesExperiment
```

### Background

Extended auto-correlation are characteristic of a process that display trend. It just means that at a given point in time, the data will be highly correlated to points in the near history as well as further back in history. So if a point far back in history is above the mean, the current point will also likely be above the mean as well. Hence the trend is maintained for a long time. 

```python
#### Get the data ---
y = get_data("1", folder="time_series/random_walk")

#### Setup the experiment ----
exp = TimeSeriesExperiment()
exp.setup(data=y, fh=12, seasonal_period=12, session_id=42)

#### Plot the ACF for the dataset ----
# `plot_model` call without an estimator works on the original dataset
exp.plot_model(plot="acf")
```

![image](https://user-images.githubusercontent.com/33585645/138612788-975f7aab-3f97-407f-b2d5-abb2fc2aee3b.png)

### Observations
- The data shows extended auto-correlation. The ACF remains positive and above the significance level all the way from lag1 to beyond lag 15. This is a classic symptom of trend in the data set.
- **Interactive plots make it easy for users to hover over the plot to get the exact values at any lag, making it easier to analyze the data.**

### Background

Repeated spikes at harmonics (multiples) of a certain time period indicate seasonality in the data. Seasonality means that the data tends to repeat after a certain number of observations. An example of seasonal data is sales of winter clothes in a retail store. This usually peaks right before the winter months and dies down in summer months. Hence it exhibits a seasonal pattern. The sales in winter this year will be highly correlated to the sales in winter of last year and the year before that. If this data is collected monthly, we will see spikes in ACF at lags of 12 (one year back), 24 (two years back), 36 (three years back), etc.

```python
#### Get the data ---
y = get_data("114", folder="time_series/seasonal")

#### Setup the experiment ----
exp = TimeSeriesExperiment()
exp.setup(data=y, fh=12, seasonal_period=12, session_id=42)

#### Plot the ACF for the dataset ----
# `plot_model` call without an estimator works on the original dataset
exp.plot_model(plot="acf")
```

![image](https://user-images.githubusercontent.com/33585645/138612273-38f7af4f-854c-444e-a722-fd2158ab0a6c.png)

### Observations
- The peaking in the ACF at harmonics of 12 (12, 24, 36 ...) indicate that the seasonality of 12 is present in the data set.
- **Interactive plots make it easy for users to hover over the plot to get the exact values at any lag, making it easier to analyze the data.**

### Background

ACF where the magnitude is not significant at any lag is indicative of white noise. i.e. there is no signal in the data that can be modeled. In other words, there is no temporal signal that can be derived from any lagged values of the dataset. It is best to model this kind of time series dataset with a `naive` (also known as Grand Means) model where the prediction is simply the mean value of the data used to train the model. This is the "best" estimate of any future prediction when the data is white noise.

```python
#### Get the data ---
y = get_data("1", folder="time_series/white_noise")

#### Setup the experiment ----
exp = TimeSeriesExperiment()
exp.setup(data=y, fh=12, seasonal_period=12, session_id=42)

#### Plot the ACF for the dataset ----
# `plot_model` call without an estimator works on the original dataset
exp.plot_model(plot="acf")
```

![image](https://user-images.githubusercontent.com/33585645/138612378-d908d1c8-9aaa-4674-9e62-2cd1bf5b399f.png)

### Observations
- The fact that no ACF values are above the significance level is indicative of the fact that this data is consistent with white noise. 
- NOTE: The ACF value at a lag of 0 is always 1 since a data set is always perfectly correlated to itself. 

# Suggested Next Reads

- https://github.com/pycaret/pycaret/discussions/1800

**Comment by moezali1** [comment]
PyCaret 4.0 sunset the legacy time-series module (sktime/prophet integration), so this FAQ no longer maps to a supported code path.

If you still need time-series forecasting, use sktime / prophet / statsforecast directly — those projects have surpassed what the PyCaret wrapper offered.

For 4.0 docs, see https://pycaret.org/docs and the cleanup notice at https://github.com/pycaret/pycaret/issues/4177.

Closing as outdated.
